azurelinuxagent/ga/exthandlers.py

```diff
@@ -935,11 +976,23 @@ def handle_uninstall(self, ext_handler_i, extension):
                 ahead and remove all of them at once if HandlerState==Uninstall.
                 CRP will only set the HandlerState to Uninstall if all its extensions are set to be disabled)
             2- Finally uninstall the handler
+
+        Before running any extension code, check extension policy. If extension is disallowed by policy, a PolicyError will be raised and uninstall will be blocked.
+        Note: if the extension was never installed, uninstall is allowed even for disallowed extensions because no extension code will be executed.
         """
         handler_state = ext_handler_i.get_handler_state()
         ext_handler_i.logger.info("[Uninstall] current handler state is: {0}", handler_state.lower())
         if handler_state != ExtHandlerState.NotInstalled:
+            # If extension is installed, check policy and raise an error if the extension is disallowed (e.g., not in allowlist, signature not previously validated when required).
+            # Uninstall extension goal states do not include encoded signature, so if the extension signature was previously validated, we treat it as signed.
+            ext_dir = ext_handler_i.get_base_dir()
+            extension_is_signed = os.path.exists(ext_dir) and signature_has_been_validated(ext_dir)
+            if self._policy_engine is None:
```

Comment:
allowing self._policy_engine to be None requires adding these checks all over the code; eventually somebody is going to forget to check, code may be moved to a place without protection, etc.